`for _, mut j in map` doesn't work

**V version:** 
master

**OS:** 
win10

**What did you do?**
```V
fn main() {
	mut m := {'foo': 1, 'bar': 2}
	for _, mut j in m {
		j = 3
	}
	println(m)
}
```

**What did you expect to see?**
{'foo': 3, 'bar': 3}
 
**What did you see instead?**
{'foo': 1, 'bar': 2}

More complicated cases would cause C error, e.g.
```V
fn main() {
	mut m := {'foo': {'a': 1}, 'bar': {'b': 2}}
	for _, mut j in m {
		j = {'c': 10}
	}
	println(m)
}
```

```
C:\Users\Roy\AppData\Local\Temp\v\sample.18358481528828516735.tmp.c: In function 'main__main':
C:\Users\Roy\AppData\Local\Temp\v\sample.18358481528828516735.tmp.c:8562:5: error: incompatible types when assigning to type 'map_string_int *' {aka 'struct map *'} from type 'map' {aka 'struct map'}
   j = new_map_init_2(&map_hash_string, &map_eq_string, &map_clone_string, &map_free_string, 1, sizeof(string), sizeof(int), _MOV((string[1]){_SLIT("c"), }), _MOV((int[1]){10, }));
     ^
builder error:
==================
C error. This should never happen.
```

--- a/vlib/v/gen/cgen.v
+++ b/vlib/v/gen/cgen.v
@@ -1362,11 +1362,21 @@ fn (mut g Gen) for_in(it ast.ForInStmt) {
 				g.write(' = (*(voidptr*)')
 			} else {
 				val_styp := g.typ(it.val_type)
-				g.write('\t$val_styp ${c_name(it.val_var)} = (*($val_styp*)')
+				if it.val_type.is_ptr() {
+					g.write('\t$val_styp ${c_name(it.val_var)} = &(*($val_styp)')
+				} else {
+					g.write('\t$val_styp ${c_name(it.val_var)} = (*($val_styp*)')
+				}
 			}
 			g.writeln('DenseArray_value(&$atmp${arw_or_pt}key_values, $idx));')
 		}
+		if it.val_is_mut {
+			g.for_in_mul_val_name = it.val_var
+		}
 		g.stmts(it.stmts)
+		if it.val_is_mut {
+			g.for_in_mul_val_name = ''
+		}
 		if it.key_type == table.string_type && !g.is_builtin_mod {
 			// g.writeln('string_free(&$key);')
 		}

--- a/vlib/v/tests/for_in_mut_val_test.v
+++ b/vlib/v/tests/for_in_mut_val_test.v
@@ -49,3 +49,12 @@ fn test_for_in_mut_val_of_map() {
 	println(m)
 	assert '$m' == "{'hello': [2, 4, 6]}"
 }
+
+fn test_for_in_mut_val_of_map_direct() {
+	mut m := {'foo': 1, 'bar': 2}
+	for _, mut j in m {
+		j = 3
+	}
+	println(m)
+	assert '$m' == "{'foo': 3, 'bar': 3}"
+}

